OSDOCS#13057: Cluster-wide architecture preferred/weighted affinity (#89119)
openshift/openshift-docs: modules/multi-arch-creating-podplacment-config.adoc

  namespaceSelector: <3>
    matchExpressions:
      - key: multiarch.openshift.io/exclude-pod-placement 
        operator: DoesNotExist

>   plugins:
>     nodeAffinityScoring: <4>
>       enabled: true <5>
>       platforms: <6>
>           - architecture: amd64 <7>
>             weight: 100 <8>
>           - architecture: arm64
>             weight: 50

Reviewer: ```suggestion
  plugins:
    nodeAffinityScoring: <4>
      enabled: true <5>
      platforms: <6>
        - architecture: amd64 <7>
          weight: 100 <8>
        - architecture: arm64
          weight: 50
```

the currect indentation is no problem for running the command, to keep the format consistent with above `namespaceSelector` field, hope to change like above.

Author: Updated as per the suggestion.

----
<1> You must set this field value to `cluster`. 
<2> Optional: You can set the field value to `Normal`, `Debug`, `Trace`, or `TraceAll`. The value is set to `Normal` by default. 
<3> Optional: You can configure the `namespaceSelector` to select the namespaces in which the Multiarch Tuning Operator's pod placement operand must process the `nodeAffinity` of the pods. All namespaces are considered by default.

> <4> Optional: This plugin allows you to set architecture preferences for pod placement. If you enable this plugin, the scheduler prioritizes nodes for pod placement based on the scores configured for architectures in the `nodeAffinityScoring.platforms` field.

Reviewer 2: ```suggestion
<4> Optional: You can use this plugin to set architecture preferences for pod placement. If you enable this plugin, the scheduler prioritizes nodes for pod placement based on the scores configured for architectures in the `nodeAffinityScoring.platforms` field.
```

Avoid using "allow": https://www.ibm.com/docs/en/ibm-style?topic=word-usage#a

> Avoid stating that inanimate objects grant abilities to people, as in “the product allows you to…”. Whenever possible, use a direct, user-focused alternative such as “you can use the product to…” or “with this product, users can…”.

<5> Optional: To enable the `nodeAffinityScoring` plugin, set this field to `true`. This field is set to `false` by default.
<6> Optional: Defines the list of architectures and their corresponding scores, influencing pod placement priority. 

> <7> Specify the node architecture to be considered for pod placement. The scheduler assigns a preference based on the score configured for this architecture.

Reviewer: the architecture can be set to `arm64`, `amd64`,  `ppc64le`, or `s390x`, can we also add this info to the doc?

Author: Updated as per the suggestion.

  namespaceSelector: <3>
    matchExpressions:
      - key: multiarch.openshift.io/exclude-pod-placement 
        operator: DoesNotExist
  plugins: <4>
    nodeAffinityScoring: <5>
      enabled: true <6>
      platforms: <7>
        - architecture: amd64 <8>
          weight: 100 <9>
        - architecture: arm64
          weight: 50
----
<1> You must set this field value to `cluster`. 
<2> Optional: You can set the field value to `Normal`, `Debug`, `Trace`, or `TraceAll`. The value is set to `Normal` by default. 
<3> Optional: You can configure the `namespaceSelector` to select the namespaces in which the Multiarch Tuning Operator's pod placement operand must process the `nodeAffinity` of the pods. All namespaces are considered by default.
<4> Optional: Includes a list of plugins for architecture-aware workload scheduling.

> <5> Optional: You can use this plugin to set architecture preferences for pod placement. If you enable this plugin, the scheduler prioritizes nodes for pod placement based on the scores configured for architectures in the `nodeAffinityScoring.platforms` field.

Reviewer 3: that can run the pod as in that the scheduler prioritizes compatible-in-arch nodes based on the score. Non-compatible-in-architecture nodes are discarded before the nodeAffinityScoring plugin takes action.

> <5> Optional: You can use this plugin to set architecture preferences for pod placement. If you enable this plugin, the scheduler prioritizes nodes for pod placement based on the scores configured for architectures in the `nodeAffinityScoring.platforms` field.

Reviewer 3: Users can set scores for arm64, amd64, ppc64le, and s390x.

For pods compatible with amd64 and ppc64le, the scheduler will prioritize among amd64 and ppc64le nodes based on the scores set by the users in the CPPC.

> <5> Optional: You can use this plugin to set architecture preferences for pod placement. If you enable this plugin, the scheduler prioritizes nodes for pod placement based on the scores configured for architectures in the `nodeAffinityScoring.platforms` field.

Author: Got it!
Can we update it as follows:
You can use this plugin to set architecture preferences for pod placement. If you enable this plugin, the scheduler first filters out nodes that do not meet the pod’s requirements. Then, it prioritizes the remaining nodes based on the architecture scores defined in the `nodeAffinityScoring.platforms` field.

> <5> Optional: You can use this plugin to set architecture preferences for pod placement. If you enable this plugin, the scheduler first filters out nodes that do not meet the pod’s requirements. Then, it prioritizes the remaining nodes based on the architecture scores defined in the `nodeAffinityScoring.platforms` field.

Reviewer 4: ```suggestion
<5> Optional: You can use this plugin to set architecture preferences for pod placement. When enabled, the scheduler first filters out nodes that do not meet the pod’s requirements. Then, it prioritizes the remaining nodes based on the architecture scores defined in the `nodeAffinityScoring.platforms` field.
```

Author: Updated as per the suggestion.

> <6> Optional: To enable the `nodeAffinityScoring` plugin, set this field to `true`. This field is set to `false` by default.

Reviewer 4: ```suggestion
<6> Optional:  Set this field to `true` to enable the `nodeAffinityScoring` plugin. The default value is `false`.
```

> <7> Optional: Defines the list of architectures and their corresponding scores, influencing pod placement priority. 

Reviewer 3: ```suggestion
<7> Optional: Defines the list of architectures and their corresponding scores for the scheduler node affinity scoring plugin.
```

Reviewer 3: Let's try to keep on the score/scoring terminology where possible. Priority is used elsewhere and also in other contexts that can confuse some users.

Author: Sure! I suggest that we update it as follows:
"Defines a list of architectures and their corresponding scores."
We've already provided the background in the desc for the preceding fields, and also the fields after this field.

> <8> Specify the node architecture to be considered for pod placement. The scheduler assigns a preference based on the score configured for this architecture. You can set this field to `arm64`, `amd64`, `ppc64le`, or `s390x`.

Reviewer 3: ```suggestion
<8> Specify the node architecture to score. The scheduler will prioritize nodes available to host a pod based on the scores set for their architecture, among others set in the pods specifications or other configurations. You can set this field to `arm64`, `amd64`, `ppc64le`, or `s390x`.
```

Author: Updated as follows:

Specify the node architecture to score. The scheduler prioritizes nodes for pod placement based on the architecture scores you set and the scheduling requirements defined in the pod specification. You can set this field to `arm64`, `amd64`, `ppc64le`, or `s390x`.

> <8> Specify the node architecture to score. The scheduler prioritizes nodes for pod placement based on the architecture scores that you set and the scheduling requirements defined in the pod specification. You can set this field to `arm64`, `amd64`, `ppc64le`, or `s390x`.

Reviewer 4: ```suggestion
<8> Specifies the node architecture to score. The scheduler prioritizes nodes for pod placement based on the architecture scores that you set and the scheduling requirements defined in the pod specification. Accepted values are `arm64`, `amd64`, `ppc64le`, or `s390x`.
```

Reviewer 4: When you say `score` do you mean the `weight` spec? I think you do and if so I'd clarify this somewhere for the user.

Author: "Specifies" is apt as the users must be doing it.

> <9> Assign a score to the architecture. The value for this field must be configured in the range of `1` (lowest priority) to `100` (highest priority). The scheduler uses this score to influence pod placement, favoring nodes with architectures that have higher scores.

Reviewer 4: Maybe, assign rather than `influence`.

    matchExpressions:
      - key: multiarch.openshift.io/exclude-pod-placement 
        operator: DoesNotExist 

>   nodeAffinityScoring: <4>
>     enabled: true <5>
>     platforms: <6>
>         - architecture: amd64 <7>
>           weight: 100 <8>
>         - architecture: arm64
>           weight: 50

Reviewer 3: ```suggestion
  plugins:
    nodeAffinityScoring: <4>
      enabled: true <5>
      platforms: <6>
          - architecture: amd64 <7>
            weight: 100 <8>
          - architecture: arm64
            weight: 50
```

Reviewer 3: cc @AnnaZivkovic  @pkenchap, is this correct?

Reviewer 5: Yes , this is correct . under specs we need plugins . 

```
spec:
  plugins:
    nodeAffinityScoring:
        enabled: true
        platforms:
```

Reviewer 6: That is correct

Author: Updated as per the suggestion.

----
<1> You must set this field value to `cluster`. 
<2> Optional: You can set the field value to `Normal`, `Debug`, `Trace`, or `TraceAll`. The value is set to `Normal` by default. 
<3> Optional: You can configure the `namespaceSelector` to select the namespaces in which the Multiarch Tuning Operator's pod placement operand must process the `nodeAffinity` of the pods. All namespaces are considered by default.

> <4> Optional: This plugin allows you to set architecture preferences for pod placement. If you enable this plugin, the scheduler prioritizes nodes for pod placement based on the scores configured for architectures in the `nodeAffinityScoring.platforms` field.

Reviewer 3: Can we add a Plugins subsection with one paragraph describing the nodeAffinityScoring plugin?

Reviewer 6: This will be especially useful as we add more plugins.

Author: I agree! Adding a subsection will be useful.
However, I think it will be best to create a subsection for the plugins in the future when we have multiple plugins. That way we can avoid stuffing all the info in the configuration section and add it in the plugins subsection instead.
For now, we have added all the info about the new plugin here in the config section. This should suffice for now.
WDYT? @AnnaZivkovic @aleskandro

Reviewer 3: Up to you to do this now, or in the next version.
